corehq/apps/accounting/tests/test_customer_invoicing.py

```diff
@@ -0,0 +1,389 @@
+
+        num_users_advanced = self.advanced_rate.monthly_limit + 1
+        generator.arbitrary_commcare_users_for_domain(self.domain2.name, num_users_advanced)
+
+        invoice_date = utils.months_from_date(self.subscription.date_start,
+                                              random.randint(2, self.subscription_length))
+        invoice_start, invoice_end = get_previous_month_date_range(invoice_date)
+        invoice_factory = CustomerAccountInvoiceFactory(
+            account=self.account,
+            date_start=invoice_start,
+            date_end=invoice_end
+        )
+        invoice_factory.create_invoice()
+        invoice_set = self.subscription.invoice_set.union(self.sub2.invoice_set.all(), self.sub3.invoice_set.all())
+        self.assertEqual(invoice_set.count(), 1)
+
+        invoice = invoice_set.first()
+        user_line_items = invoice.lineitem_set.get_feature_by_type(FeatureType.USER)
+        self.assertEqual(user_line_items.count(), 2)
+        for user_line_item in user_line_items:
+            self.assertIsNone(user_line_item.base_description)
+            self.assertEqual(user_line_item.base_cost, Decimal('0.0000'))
+            num_to_charge = num_users - self.user_rate.monthly_limit
+            self.assertEqual(num_to_charge, user_line_item.quantity)
+            if self.user_rate.feature.name == user_line_item.feature_rate.feature.name:
+                self.assertEqual(user_line_item.unit_cost, self.user_rate.per_excess_fee)
+                self.assertEqual(user_line_item.total, self.user_rate.per_excess_fee * num_to_charge)
+                self.assertEqual(user_line_item.subtotal, self.user_rate.per_excess_fee * num_to_charge)
+            elif user_line_item.feature_rate.feature.name == self.advanced_rate.feature.name:
+                self.assertEqual(user_line_item.unit_cost, self.advanced_rate.per_excess_fee)
+                self.assertEqual(user_line_item.total, self.advanced_rate.per_excess_fee * num_to_charge)
+                self.assertEqual(user_line_item.subtotal, self.advanced_rate.per_excess_fee * num_to_charge)
+
+
+class TestSmsLineItem(BaseCustomerInvoiceCase):
+
+    def setUp(self):
+        super(TestSmsLineItem, self).setUp()
+        self.sms_rate = self.subscription.plan_version.feature_rates.filter(
+            feature__feature_type=FeatureType.SMS
+        ).get()
+        self.advanced_rate = self.advanced_plan.feature_rates.filter(feature__feature_type=FeatureType.SMS).get()
+        self.invoice_date = utils.months_from_date(
+            self.subscription.date_start, random.randint(2, self.subscription_length)
+        )
+        self.sms_date = utils.months_from_date(self.invoice_date, -1)
+
+    def tearDown(self):
+        self._delete_sms_billables()
+        super(TestSmsLineItem, self).tearDown()
+
+    def test_under_limit(self):
+        num_sms = random.randint(0, self.sms_rate.monthly_limit // 2)
```

Comment:
I'd recommend replacing ```0``` with a positive number, so you're always testing for at least one billable.  Otherwise, if someone wrote could that charged for every SMS, this test case would sometimes pass even though the code is broken.